examples/platform/silabs/SiWx917/SiWx917/sl_wifi_if.c

```diff
@@ -655,14 +721,169 @@ static sl_status_t wfx_rsi_do_join(void)
                 wfx_retry_interval_handler(is_wifi_disconnection_event, wfx_rsi.join_retries);
                 if (is_wifi_disconnection_event || wfx_rsi.join_retries <= MAX_JOIN_RETRIES_COUNT)
                 {
-                    xEventGroupSetBits(wfx_rsi.events, WFX_EVT_STA_START_JOIN);
+                    event.eventType = WFX_EVT_STA_START_JOIN;
+                    WfxPostEvent(&event);
                 }
             }
         }
     }
     return status;
 }
 
+void HandleDHCPPolling()
+{
+    struct netif * sta_netif;
+    WfxEvent_t event;
+
+    sta_netif = wfx_get_netif(SL_WFX_STA_INTERFACE);
+#if (CHIP_DEVICE_CONFIG_ENABLE_IPV4)
+    uint8_t dhcp_state = dhcpclient_poll(sta_netif);
+    if (dhcp_state == DHCP_ADDRESS_ASSIGNED && !hasNotifiedIPV4)
+    {
+        wfx_dhcp_got_ipv4((uint32_t) sta_netif->ip_addr.u_addr.ip4.addr);
+        hasNotifiedIPV4 = true;
+        if (!hasNotifiedWifiConnectivity)
+        {
+            wfx_connected_notify(CONNECTION_STATUS_SUCCESS, &wfx_rsi.ap_mac);
+            hasNotifiedWifiConnectivity = true;
+        }
+    }
+    else if (dhcp_state == DHCP_OFF)
+    {
+        wfx_ip_changed_notify(IP_STATUS_FAIL);
+        hasNotifiedIPV4 = false;
+    }
+#endif /* CHIP_DEVICE_CONFIG_ENABLE_IPV4 */
+    /* Checks if the assigned IPv6 address is preferred by evaluating
+     * the first block of IPv6 address ( block 0)
+     */
+    if ((ip6_addr_ispreferred(netif_ip6_addr_state(sta_netif, 0))) && !hasNotifiedIPV6)
+    {
+        wfx_ipv6_notify(GET_IPV6_SUCCESS);
+        hasNotifiedIPV6 = true;
+        event.eventType = WFX_EVT_STA_DHCP_DONE;
+        WfxPostEvent(&event);
+        if (!hasNotifiedWifiConnectivity)
+        {
+            wfx_connected_notify(CONNECTION_STATUS_SUCCESS, &wfx_rsi.ap_mac);
+            hasNotifiedWifiConnectivity = true;
+        }
+    }
+}
+
+void WfxPostEvent(WfxEvent_t * event)
+{
+    sl_status_t status = osMessageQueuePut(sWifiEventQueue, event, 0, 0);
+
+    if (status != osOK)
+    {
+        SILABS_LOG("WfxPostEvent: failed to post event with status: %d", status);
+    }
+}
+
+void ProcessEvent(WfxEvent_t inEvent)
```

Comment:
There's a repeated block of code in ProcessEvent for setting hasNotifiedWifiConnectivity to false, resetting IP notification flags, and calling wfx_lwip_set_sta_link_down  for both WFX_EVT_STA_DISCONN and some error cases in the WFX_EVT_DHCP_POLL (like dhcp_state being DHCP_OFF). You could potentially create a helper function to encapsulate this logic to improve readability and maintainability.

Reply:
Adde helpers in Process event and HandleDHCP